README.md
# CowTranslator.js 🐄

JavaScript library to translate cow language 🐄 to text

![](assets/banner.png)

# jsDelivr 

```html
<script src="https://cdn.jsdelivr.net/gh/SkwalExe/cowTranslator.js@v1.1.0/dist/cow-translator.min.js"></script>
```

# NPM module 

Install the npm module 

```bash
npm install cow-translator
```

And import it in your project 

```js
const cowTranslator = require('cow-translator');
```

# Setting up 

You can import the library into you website with [JsDelivr](#JsDelivr) or, you can use the [npm module](#NPM-module) and import it in your project.

# Usage 📝

This library provides 2 functions:
- `cowTranslator.cowToText` : translate cow language to text
- `cowTranslator.textToCow` : translate text to cow language

Each of these functions returns a translationResult class : 

- `translationResult.text` : the human version
- `translationResult.cow` : the cow version
- `translationResult.error` : the error message if any
- `translationResult.success` : whether the translation was successful or not
- `translationResult.warning` : whether warnings were generated or not

## Human to cow

**Translate Hello World ! to cow language**

```js
let cow = cowTranslator.textToCow("Hello world !");
```

`.cow` is the cow version of the text

```js
cow.cow // "mOoOoOO moooOoo mooOoOO mooOoOO mooOOOo mOOOOOo moOoOOo mooOOOo moOoooO mooOoOO mooooOO mOOOOOo MoooooO"
```

The translation is successful

```js
cow.success // true
```

## Cow to human

**Translate `"mOoOoOO moooOoo mooOoOO mooOoOO mooOOOo mOOOOOo"` to human language**

```js
let human = cowTranslator.cowToText("mOoOoOO moooOoo mooOoOO mooOoOO mooOOOo mOOOOOo");
```

`.text` is the human version

```js
human.text // "Hello"
```

## Errors and warnings

If an error occured and the translation was not successful, the following properties are set:

- `translationResult.success` : ⛔ false
- `translationResult.error` : the error message
  
If warnings were generated, the following properties are set:

- `translationResult.warning` : ⚠️ true
- `translationResult.error` : the warning message
- `translationResult.success` : ✅ true


# final

If you have any problem, don't hesitate to open an issue

# Contributing

1. Start by [**forking** this repository](https://github.com/SkwalExe/cowTranslator.js/fork)

2. Then clone your fork to your local machine.
  ```git
  git clone https://github.com/your-username/cowTranslator.js.git
  ```

3. Install dev dependencies
```npm
npm install --save-dev
```

4. Create a new branch
  ```git
  git checkout -b super-cool-feature
  ```

5. Then make your changes

6. Update the changelog and version number if needed (using [Semantic Versioning](https://semver.org)) 
  ```bash
  # bug fix
  npm version patch --no-git-tag-version

  # add a new feature 
  npm version minor --no-git-tag-version
  
  # changes that break backwards compatibility
  npm version major --no-git-tag-version
  ```

7. List and correct linting errors
  ```bash
  npm run lint
  ```

8. Update the minified/browser version of the library
  ```bash
  npm run build
  ```


9. Once you're done, commit your changes and push them to the remote repository.
  ```git
  git add --all
  git commit -m "Add super-cool-feature"
  git push origin super-cool-feature
  ```

10. Then, open a pull request on GitHub from your fork.
    1. Go to [this link](https://github.com/SkwalExe/cowTranslator.js/compare/)
    2. Click compare across forks
    3. On the right, on `head repository` select your fork
    4. And on `compare` select the branch you just created
    5. Click on `Create Pull Request` and submit your pull request

<a href="https://github.com/SkwalExe#ukraine"><img src="https://raw.githubusercontent.com/SkwalExe/SkwalExe/main/ukraine.jpg" width="100%" height="15px" /></a>

CHANGELOG.md
# 1.1.0
- Change project structure

# 1.0.0
- Translation result returned as a class with more informations
- all functions are gathered in a single class

# 0.1.0
- Initial release

src/index.js
let CowTranslator = {

  TranslationResult: class {
    text = ''; // Human language
    cow = ''; // Cow language
    error = ''; // Error / warning message
    warning = false; // If any warning was found
    success = false; // If the translation was successful
  },

  SUPPORTED_CHARS: [ 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'cV', 'W', 'X', 'Y', 'Z', ' ', '.', ',', '!', '?', ':', ';', '"', '\'', '`', '~', '\\', '/', '|', '=', '-', '_', '+', '(', ')', '[', ']', '{', '}', '\n', '\t', '\r' ],

  MOO: 'moooooo',


  /*
   * This function will
   *
   * convert a string to a binary string (uppercase : 1; lowercase : 0)
   * => "mooOoOo"
   *        ↓ ↓
   * => "0001010"
   */
  stringToBinString: function(input) {
    let output = '';

    input.split('').forEach((char) => {
      output += char.toUpperCase() === char ? '1' : '0';
      /*
       * Char.toUpperCase() === char
       * checks if the char is uppercase
       */
    })
    return output;
  },

  /*
   * This function will
   *
   * convert a binary string to a decimal number
   * => "0001010"
   * => 10
   */
  binStringToDecimal: function(input) {
    return parseInt(input, 2);
  },


  /*
   * This function will :
   *
   * step 1 - convert a string to a binary string (uppercase : 1; lowercase : 0)
   * => "mooOoOo"
   *        ↓ ↓
   * => "0001010"
   *
   * step 2 - Then convert the binary string to a decimal number
   * => "0001010"
   * => 10
   *
   * step 3 - add the specified number to the decimal number
   * => 10 + 5
   * => 15
   *
   * step 4 - convert back the decimal number to a binary string
   * => 15
   * => "0001111"
   *
   * step 5 - and finally convert the binary string to the origin string but with the specified upper/lowercase
   * => "0001111"
   *        ↓↓↓↓
   * => "mooOOOO"
   */
  addToStringLikeBinary: function(input, toAdd) {
    let output = '';
    let inputAsBin;
    let inputAsDecimal;
    let outputAsBin;

    // Step 1

    inputAsBin = this.stringToBinString(input);

    // Step 2

    inputAsDecimal = this.binStringToDecimal(inputAsBin);

    // Step 3

    inputAsDecimal += toAdd;

    // Step 4

    outputAsBin = inputAsDecimal.toString(2);

    while (outputAsBin.length < inputAsBin.length) {
      outputAsBin = '0' + outputAsBin;
    }
    // Step 5
    for (let i = 0; i < outputAsBin.length; i++) {
      let char = input[i];
      let currentBit = outputAsBin[i];

      if (currentBit === '1') {
        output += char.toUpperCase();
      } else {
        output += char.toLowerCase();
      }
    }

    return output;
  },


  textToCow: function(input) {

    let result = new this.TranslationResult();
    result.text = input;
    if (!input) {
      console.warn('No text specified');
      result.error = 'No text specified';

      return result;
    }

    input = input.trim();


    let inputAsNumbers = '';
    let errors = 0;

    input.split('').forEach(function(char) {

      if (this.SUPPORTED_CHARS.indexOf(char) === -1) {
        console.warn('Unsupported character: ' + char);
        result.warning = true;
        result.error = 'Unsupported character: ' + char;
        errors++;
        return;
      }

      let charAsNumber = (this.SUPPORTED_CHARS.indexOf(char) >= 10 ? '' : '0') + this.SUPPORTED_CHARS.indexOf(char);

      inputAsNumbers += charAsNumber;
    }, this)

    if (errors === input.length) {
      result.error = 'No supperted characters found';
      return result;
    }


    let inputAsMoos = '';

    inputAsNumbers.match(/.{1,2}/g).forEach(function(number) {
      let numberAsMoo = this.addToStringLikeBinary(this.MOO, parseInt(number));
      inputAsMoos += numberAsMoo + ' ';
    }, this)


    result.cow = inputAsMoos.trim();
    result.success = true;

    return result;
  },

  cowToText: function(input) {


    let result = new this.TranslationResult();
    result.cow = input;

    if (!input) {
      console.warn('No text specified');
      result.error = 'No text specified';

      return result;
    }

    input = input.trim();


    let moosVec = input.split(' ');

    let inputAsText = '';

    let errors = 0;

    moosVec.forEach(moo => {
      if (moo.toLowerCase() !== this.MOO) {
        errors++;
        console.warn('Invalid moo: ' + moo);
        result.warning = true;
        result.error = 'Invalid moo: ' + moo;

        return;
      }

      let mooAsBin = this.stringToBinString(moo);
      let mooAsDecimal = this.binStringToDecimal(mooAsBin);

      if (this.SUPPORTED_CHARS[mooAsDecimal] === undefined) {
        errors++;
        console.warn('Unsupported char : ' + moo);
        result.warning = true;
        result.error = 'Unsupported char : ' + moo;

        return;
      }

      inputAsText += this.SUPPORTED_CHARS[mooAsDecimal];

    }, this);

    if (errors === moosVec.length) {
      result.error = 'No supperted characters found';
      return result;

    }
    result.text = inputAsText;
    result.success = true;


    return result;
  }
}


if (typeof module !== 'undefined') {
  module.exports = CowTranslator;
}


src/cowTranslator.js
let CowTranslator = {

    TranslationResult: class {
        text = ""; // human language
        cow = ""; // cow language
        error = ""; // error / warning message
        warning = false; // if any warning was found
        success = false; // if the translation was successful
    },

    SUPPORTED_CHARS: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r',
        's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
        'S', 'T', 'U', 'cV', 'W', 'X', 'Y', 'Z', ' ', '.', ',', '!', '?', ':', ';', '"', '\'', '`',
        '~', '\\', '/', '|', '=', '-', '_', '+', '(', ')', '[', ']', '{', '}', '\n', '\t', '\r',
    ],

    MOO: "moooooo",


    // This function will
    //
    // convert a string to a binary string (uppercase : 1; lowercase : 0)
    // => "mooOoOo"
    //        ↓ ↓
    // => "0001010"
    stringToBinString: function(input) {
        var output = "";

        input.split("").forEach(function(char) {
            output += char.toUpperCase() == char ? "1" : "0";
            // char.toUpperCase() == char
            // checks if the char is uppercase
        })
        return output;
    },

    // This function will
    //
    // convert a binary string to a decimal number
    // => "0001010"
    // => 10
    binStringToDecimal: function(input) {
        return parseInt(input, 2);
    },


    // This function will :
    //
    // step 1 - convert a string to a binary string (uppercase : 1; lowercase : 0)
    // => "mooOoOo"
    //        ↓ ↓
    // => "0001010"
    //
    // step 2 - Then convert the binary string to a decimal number
    // => "0001010"
    // => 10
    //
    // step 3 - add the specified number to the decimal number
    // => 10 + 5
    // => 15
    //
    // step 4 - convert back the decimal number to a binary string
    // => 15
    // => "0001111"
    //
    // step 5 - and finally convert the binary string to the origin string but with the specified upper/lowercase
    // => "0001111"
    //        ↓↓↓↓
    // => "mooOOOO"
    addToStringLikeBinary: function(input, toAdd) {
        var output = "";
        var inputAsBin;
        var inputAsDecimal;
        var outputAsBin;

        // step 1

        inputAsBin = this.stringToBinString(input);

        // step 2

        inputAsDecimal = this.binStringToDecimal(inputAsBin);

        // step 3

        inputAsDecimal += toAdd;

        // step 4

        outputAsBin = inputAsDecimal.toString(2);

        while (outputAsBin.length < inputAsBin.length) {
            outputAsBin = "0" + outputAsBin;
        }
        // step 5
        for (var i = 0; i < outputAsBin.length; i++) {
            var char = input[i];
            var currentBit = outputAsBin[i];

            if (currentBit == "1") {
                output += char.toUpperCase();
            } else {
                output += char.toLowerCase();
            }
        }

        return output;
    },



    textToCow: function(input) {

        var result = new this.TranslationResult();
        result.text = input;
        if (!input) {
            console.warn("No text specified");
            result.error = "No text specified";

            return result;
        }

        input = input.trim();


        var inputAsNumbers = "";
        var errors = 0;

        input.split("").forEach(function(char) {

            if (this.SUPPORTED_CHARS.indexOf(char) == -1) {
                console.warn("Unsupported character: " + char);
                result.warning = true;
                result.error = "Unsupported character: " + char;
                errors++;
                return;
            }

            var charAsNumber = (this.SUPPORTED_CHARS.indexOf(char) >= 10 ? "" : "0") + this.SUPPORTED_CHARS.indexOf(char);

            inputAsNumbers += charAsNumber;
        }, this)

        if (errors == input.length) {
            result.error = "No supperted characters found";
            return result;
        }


        var inputAsMoos = "";

        inputAsNumbers.match(/.{1,2}/g).forEach(function(number) {
            var numberAsMoo = this.addToStringLikeBinary(this.MOO, parseInt(number));
            inputAsMoos += numberAsMoo + " ";
        }, this)


        result.cow = inputAsMoos.trim();
        result.success = true;

        return result;
    },

    cowToText: function(input) {


        var result = new this.TranslationResult();
        result.cow = input;

        if (!input) {
            console.warn("No text specified");
            result.error = "No text specified";

            return result;
        }

        input = input.trim();



        var moosVec = input.split(" ");

        var inputAsText = "";

        var errors = 0;

        moosVec.forEach(moo => {
            if (moo.toLowerCase() != this.MOO) {
                errors++;
                console.warn("Invalid moo: " + moo);
                result.warning = true;
                result.error = "Invalid moo: " + moo;

                return;
            }

            var mooAsBin = this.stringToBinString(moo);
            var mooAsDecimal = this.binStringToDecimal(mooAsBin);

            if (this.SUPPORTED_CHARS[mooAsDecimal] == undefined) {
                errors++;
                console.warn("Unsupported char : " + moo);
                result.warning = true;
                result.error = "Unsupported char : " + moo;

                return;
            }

            inputAsText += this.SUPPORTED_CHARS[mooAsDecimal];

        }, this);

        if (errors == moosVec.length) {
            result.error = "No supperted characters found";
            return result;

        }
        result.text = inputAsText;
        result.success = true;


        return result;
    }
}


if (typeof module != "undefined") {
    exports = module.exports = CowTranslator;
}